inc/features/blocks/personalization/edit.js

```diff
@@ -0,0 +1,351 @@
+ *
+ * @param {String} parentId The parent experience block client ID.
+ * @return The new fallback variant block.
+ */
+const createFallbackBlock = parentId => {
+	return createBlock( 'altis/personalization-variant', {
+		parentId,
+		audience: null,
+		fallback: true,
+	} );
+};
+
+export default compose(
+	withSelect( ( select, ownProps ) => {
+		const { clientId, attributes } = ownProps;
+		const { getBlocks } = select( 'core/block-editor' );
+
+		const parentClientId = attributes.clientId || clientId;
+		const innerBlocks = getBlocks( clientId );
+
+		// Ensure at least one variant is present as a fallback.
+		// Note TEMPLATE does not seem to have the desired effect every time.
+		if ( innerBlocks.length === 0 ) {
+			const fallbackBlock = createFallbackBlock( parentClientId );
+			innerBlocks.push( fallbackBlock );
+		} else {
+			// Add a flag to check if we have a fallback explicitly set.
+			let hasFallback = false;
+			innerBlocks.forEach( block => {
+				if ( block.attributes.fallback ) {
+					hasFallback = block.clientId;
+				}
+			} );
+			if ( ! hasFallback ) {
+				const fallbackBlock = createFallbackBlock( parentClientId );
+				innerBlocks.push( fallbackBlock );
+			}
+		}
+
+		return {
+			variants: innerBlocks,
+		};
+	} ),
+	withDispatch( ( dispatch, ownProps, registry ) => {
+		return {
+			addVariant() {
+				const { clientId, attributes } = ownProps;
+				const { replaceInnerBlocks } = dispatch( 'core/block-editor' );
+				const { getBlocks } = registry.select( 'core/block-editor' );
+
+				let innerBlocks = getBlocks( clientId );
+
+				const newVariant = createBlock( 'altis/personalization-variant', {
+					parentId: attributes.clientId,
+					audience: null,
+				} );
+
+				// Prepend the new variant.
+				innerBlocks = [
+					newVariant,
+					...innerBlocks,
+				];
+
+				// Update the inner blocks.
+				replaceInnerBlocks( clientId, innerBlocks );
+
+				// Return new client ID to enable selection.
+				return newVariant.clientId;
+			},
+			copyVariant( variantClientId ) {
+				const { replaceInnerBlocks } = dispatch( 'core/block-editor' );
+				const {
+					getBlock,
+					getBlocks,
+					getBlockRootClientId,
+				} = registry.select( 'core/block-editor' );
+
+				// Clone the the block but override the audience.
+				const fromVariant = getBlock( variantClientId );
+				const newVariant = cloneBlock( fromVariant, {
+					audience: null,
+					fallback: false,
+				} );
+
+				const experienceBlockClientId = getBlockRootClientId( variantClientId );
+				const variantBlocks = getBlocks( experienceBlockClientId );
+				const fromVariantIndex = variantBlocks.findIndex( variant => variant.clientId === variantClientId );
+
+				let innerBlocks = variantBlocks;
+
+				// If we've copied the fallback then add it just before the fallback variant.
+				// Otherwise add it just after the source block.
+				if ( fromVariant.attributes.fallback ) {
+					innerBlocks.splice( fromVariantIndex, 0, newVariant );
```

Comment:
Could use spreads instead of splicing, which avoids the need to copy:

```js
const nextBlocks = [
    ...innerBlocks.slice( 0, fromVariantIndex ),
    newVariant,
    ...innerBlocks.slice( fromVariantIndex ),
];
```
(Might be off-by-one there).

Reply:
Could do but I'm not sure why is this better?

Reply:
Splicing is a bit non-obvious in the way it's called, and it's a mutation, so can easily lead to accidentally mutating state from other places; for example here, without that `.slice` above, if `getBlocks` returned a cached value (eg) you'd end up mutating it. Spreads mean that the value remains immutable and avoids subtle bugs like that.